### DISCUSSION TITLE
debugging python code

### DISCUSSION THREAD

**Discussion by axr0284** [cocotb/cocotb#2652] [Q&A] [answered]
Hi,
 this is my first question. I am very new to cocotb. This is the only forum I could find to discuss cocotb. If there is another one more acvtive, please let me know. 
I was handed a cocotb testbench running in a docker and trying to make my way through it understanding the whole concept. I was told run make and it's gonna run the test which is does correctly
My first question is how do you actually debug the python code. Meaning is there a way to set some settings in the makefile so that I get a debugger prompt that let me run through the python code line by line and look at variables set, etc... using the make command? This is very confusing to me. Thanks for any help
Amish

**Comment by cmarqu** [comment, marked as answer]
Do these instructions help?: https://docs.cocotb.org/en/stable/troubleshooting.html?highlight=pdb#python

**Reply by axr0284** [reply]
sadly no telnet is available in the docker but this is a different problem. I will see if I can add a telnet client to it and try this. Thanks for the answer

**Reply by axr0284** [reply]
I was able to install telnet and use this. Thanks

**Comment by ktbarrett** [comment]
There is [`pdb_attach`](https://pypi.org/project/pdb-attach/) which I have used with success.

**Reply by marlonjames** [reply]
I use Visual Studio Code with [`debugpy`](https://pypi.org/project/debugpy/) for a nice debugging environment.

**Reply by axr0284** [reply]
I actually have not really tried installing cocotb manually on my computer instead of using the docker provided to me. I am running red hat 7 and using visual studio code. I might try that too. Is there any specific instructions somewhere to get cocotb to work with visual studio code?

**Reply by marlonjames** [reply]
Use this configuration in your VSCode `launch.json` file:
```json
{
  "version": "0.2.0",
  "configurations": [
      {
          "name": "Python: attach cocotb",
          "type": "python",
          "request": "attach",
          "host": "localhost",
          "port": 5678
      }
  ]
}
```
Then add something like this in your Python code where you want the debugger to break-in:
```py3
import debugpy
listen_host, listen_port = debugpy.listen(("localhost", 5678))
cocotb.log.info("Waiting for Python debugger attach on {}:{}".format(listen_host, listen_port))
# Suspend execution until debugger attaches
debugpy.wait_for_client()
# Break into debugger for user control
breakpoint()  # or debugpy.breakpoint() on 3.6 and below
```
Start up your simulation and after you see the "Waiting for Python debugger attach ..." message, start the VSCode debugging session using the configuration above.
If you are running it in a container you may have to change the host and there may be additional configuration needed to point the debugger to the right local Python files for breakpoints etc to work.

**Reply by axr0284** [reply]
sounds good. Will give it a try

**Reply by axr0284** [reply]
Ok I just tried debugpy with visual studio code and if worked off the bat. I was even able to connect to my docker container before hand. I'll add some more info for the next person who finds this
Instruction for how i set up visual studio code to use my docker container and then debug using debugpy:
I first had to install two extensions "Remote Development" and "Remote - Containers", then I connected to the docker container. After that I had to reinstall python and  pylance extensions so that it gets installed in the running docker container.
Then I created the new launch.json mentioned by marlonjames
then I ran the make command in the terminal in visual studio which started the cocotb framework and run the test. A little pop up window came up that indicated that there were new ports open. One of them was 5678.
After that I went to the run window and clicked on run Python: attach cocotb and visual studio code opened the right file and showed that the debugger was stopped at the correct line. I was then able to trace through the code normally using visual studio gui.

**Reply by hjwmedenblik** [reply]
Without knowing that this solution was available for a while, I discovered the debugpy approach too. It is a very convenient way to attach the microsoft debugpy into your VSCode environment and it allows you to use the same debug features as you are normally  used to do. I believe this should be part of the standard cocotb documentation.

**Comment by jevinskie** [comment]
I've found a way to use the pydevd remote debugger shipped with PyCharm to debug cocotb test benches.

The trick is running the testbench under PyCharm in two phases:
Phase 1: Dummy Python process that simply spawns make to re-run itself as a techbench. It exports the required pydevd information to environment variables for the real TB in phase 2 to connect to.
Phase 2: Simulator with embedded libpython running cocotb TB. Using `pydevd.settrace()`, it connects back to the PyCharm debugger, letting breakpoints in `cocotb.test`s (or anywhere else) work.

Modified adder test:
https://gist.github.com/jevinskie/86c3b759d7794f096c19d2fc1f45e7c6

You can see the meat of the changes here:
https://gist.github.com/jevinskie/86c3b759d7794f096c19d2fc1f45e7c6/revisions#diff-27c9f1b6915a4867254aaedcb508ed5a2697ee2ced9a1196cd75a771e2a634ccHR1

I'm working on getting this working under VS Code but it needs a few tweaks like debugger client authentication support.

**Comment by Jetsama** [comment]
Hi, if you are using the cocotb runners, I finally got debugging for the embedded to automatically connect! (only one session rather than two separate launch/attach) I did this is VSCode using debugpy.

For the runner, I get the debugpy port/access token and then pass them along to the subproccess using an environment variable.
```python
        try:
            from debugpy.server import cli as _debugpy_cli
            access_token = _debugpy_cli.options.adapter_access_token
            port = _debugpy_cli.options.address[1]
            os.environ["PYTHON_DEBUG_PORT_VSCODE"] = str(port)
            os.environ["PYTHON_DEBUG_ACCESS_TOKEN_VSCODE"] = access_token
        except:
            pass #no debugging today
        runner.test(
            hdl_toplevel=self.hdl_toplevel, test_module=test_module.__module__, test_args=self.build_test_args, log_file="cocotb_log.txt"#,verbose=True
        )
```
Then in my test wrapper I check if those are variables are set and connect if so. 
```python
        @functools.wraps(cls)
        async def test(top):
            logger = logging.getLogger("cocotb").getChild('CocoTB_test_wrapper')
            #SETUP DEBUGGER!
            try:
                import debugpy
                port = int(os.environ["PYTHON_DEBUG_PORT_VSCODE"])
                access_token = os.environ["PYTHON_DEBUG_ACCESS_TOKEN_VSCODE"]
                debugpy.connect(("127.0.0.1", port),access_token=access_token)
                
                print("Waiting for debugger to attach...")
                debugpy.wait_for_client()
                print("Debugger attached.")
            except:
                logger.info(f"No debugger attached")
            logger.info(f"Initiallizing testbench")
```
Here are some of the resources I found helpful:
https://github.com/microsoft/debugpy/discussions/1926
https://github.com/microsoft/debugpy/issues/552#issuecomment-787058619

Hope this is helpful!
